### PR TITLE
Pr new pmw module

### PR DESCRIPTION
**Pull request by AsiiaPine**
add pwm module for dronecan application with real pwm publishing via uavcan.equipment.esc.Status.

### CODE REVIEW

**Review by PonomarevDA — CHANGES_REQUESTED** [review]

File: Src/dronecan_application/modules/PWMModule.cpp [hunk 1]
```diff
@@ -0,0 +1,255 @@
...
+                pwm.command_val = pwm.def;
+            }
+        }
+    }
+}
+
+void PWMModule::array_command_callback(CanardRxTransfer* transfer) {
+    if (module_status != ModuleStatus::MODULE_OK) return;
+    ArrayCommand_t command;
+    int8_t res =
```
[the remark this answers is not in the document]
Comment (AsiiaPine) [reply Src/dronecan_application/modules/PWMModule.cpp]: Fixed with
```c++
void PWMModule::array_command_callback(CanardRxTransfer* transfer) {
    if (module_status != ModuleStatus::MODULE_OK) return;
    ArrayCommand_t command;
    int8_t ch_num = dronecan_equipment_actuator_arraycommand_deserialize(
        transfer, &command);
    if (ch_num == 0) {
        return;
    }
```

File: Src/dronecan_application/modules/PWMModule.cpp [hunk 2]
```diff
@@ -0,0 +1,255 @@
...
+            params[i].channel = channel;
+        } else {
+            params[i].channel = max_channel;
+            params_error = true;
+        }
+
+        auto min = paramsGetIntegerValue(params_names[i].min);
+        auto max = paramsGetIntegerValue(params_names[i].max);
+        auto def = paramsGetIntegerValue(params_names[i].def);
+        if (min <= max && def >= min && def <= max) {
```
[the remark this answers is not in the document]
Comment (AsiiaPine) [reply Src/dronecan_application/modules/PWMModule.cpp:121]: Yes, the condition checks if the parameters are ok. Thus, if min is less than max, the condition will be true and no error will occur.
```c++
        if (min <= max && def >= min && def <= max) {
            params[i].min = min;
            params[i].max = max;
            params[i].def = def;
        } else {
            params_error = true;
        }
```

[the remark this answers is not in the document]
Comment (PonomarevDA) [reply Src/dronecan_application/modules/PWMModule.cpp:121]: I mean that min < max is a valid use case. It is not an error.

Just to make it clear:
- For a ESC typically min pwm duration is 1000 us and max is 2000 us.
- For a servo it is different. We may want to rotate it in a different direction and limit the range of motion based on how we mount it in a control surface.

This is from a real usage: A single node controls 2 servos, but they move in opposite directions.

image

I can't come up with a real application, but I think it is also legal to keep def value out of the range between min and max. For example, you may want to just keep 2 states: def when disarmed and min=max when armed.

**Review by PonomarevDA — CHANGES_REQUESTED** [review]
I tested it. If all parameters are configured correctly, it works. Please, check a few notes.

File: Src/dronecan_application/modules/PWMModule.cpp [hunk 3]
```diff
@@ -0,0 +1,260 @@
...
+                publish_state = publish_esc_status;
+                // publish_state = publish_actuator_status;
+                break;
+
+            default:
+                return;
+        }
+    }
+    if (module_status == ModuleStatus::MODULE_OK) {
+        uavcanSubscribe(data_type_signature, data_type_id, callback);
```
[the remark this answers is not in the document]
Comment (AsiiaPine) [reply Src/dronecan_application/modules/PWMModule.cpp]: Done
```
void PWMModule::init() {
    logger.init("PWMModule");
    for (int i = 0; i < static_cast<uint8_t>(PwmPin::PWM_AMOUNT); i++) {
        PwmPeriphery::init(params[i].pin);
    }
    uavcanSubscribe(UAVCAN_EQUIPMENT_ESC_RAWCOMMAND_SIGNATURE, UAVCAN_EQUIPMENT_ESC_RAWCOMMAND_ID,
                                                                            raw_command_callback);
    uavcanSubscribe(UAVCAN_EQUIPMENT_ACTUATOR_ARRAY_COMMAND_SIGNATURE,
                                                        UAVCAN_EQUIPMENT_ACTUATOR_ARRAY_COMMAND_ID,
                                                        array_command_callback);
}
```

File: Src/dronecan_application/modules/PWMModule.cpp [hunk 4]
```diff
@@ -0,0 +1,260 @@
...
+    ttl_cmd = paramsGetIntegerValue(IntParamsIndexes::PARAM_PWM_CMD_TTL_MS);
+    status_pub_timeout_ms = 100;
+    uint8_t max_channel = 0;
+
+    switch (pwm_cmd_type) {
+        case 0:
+            max_channel = NUMBER_OF_RAW_CMD_CHANNELS - 1;
+            break;
+        case 1:
+            max_channel = 15;
```
[the remark this answers is not in the document]
Comment (AsiiaPine) [reply Src/dronecan_application/modules/PWMModule.cpp]: Sure, thanks
```
switch (pwm_cmd_type) {
        case 0:
            max_channel = NUMBER_OF_RAW_CMD_CHANNELS - 1;
            break;
        case 1:
            max_channel = 255;
            break;
        default:
            params_error = true;
            break;
    }
    ```

File: Src/dronecan_application/modules/PWMModule.cpp [hunk 5]
```diff
@@ -0,0 +1,260 @@
...
+                callback = array_command_callback;
+                data_type_signature =
+                    UAVCAN_EQUIPMENT_ACTUATOR_ARRAY_COMMAND_SIGNATURE;
+                data_type_id = UAVCAN_EQUIPMENT_ACTUATOR_ARRAY_COMMAND_ID;
+                publish_state = publish_esc_status;
+                // publish_state = publish_actuator_status;
+                break;
+
+            default:
+                return;
```
[the remark this answers is not in the document]
Comment (AsiiaPine) [reply Src/dronecan_application/modules/PWMModule.cpp]: Yeah, right, lets make unsubcribe function) But for now, I changed the code in the following manner:

Callback functions at the beggining checks if the command corresponds to the chosen one
```

void PWMModule::raw_command_callback(CanardRxTransfer* transfer) {
    if (module_status != ModuleStatus::MODULE_OK || pwm_cmd_type != 0) return;
    RawCommand_t command;
```
The subscription on both topics are performed at init funciton.
Publish status function works same as in the previous commits (publish_status function switch to the corresponding command type status publish function pointer during apply_params())

File: Src/dronecan_application/modules/PWMModule.cpp [hunk 6]
```diff
@@ -0,0 +1,260 @@
...
+
+        auto min = paramsGetIntegerValue(params_names[i].min);
+        auto max = paramsGetIntegerValue(params_names[i].max);
+        auto def = paramsGetIntegerValue(params_names[i].def);
+        params[i].def = def;
+        if (min == max) {
+            params_error = true;
+        } else {
+            params[i].min = max;
+            params[i].max = min;
```
[the remark this answers is not in the document]
Comment (AsiiaPine) [reply Src/dronecan_application/modules/PWMModule.cpp]: Yeah, typo from a rush) Thanks
```
        params[i].min = min;
        params[i].max = max;
```

File: Src/dronecan_application/params.yaml [hunk 7]
```diff
@@ -13,3 +13,210 @@ system.name:
...
+  max: 255
+
+pwm.1_min:
+  type: Integer
+  note: PWM duration when setpoint is min (RawCommand is 0 or Command is -1.0)
+  enum: PARAM_PWM_1_MIN
+  flags: mutable
+  default: 0
+  min: 0
+  max: 2000
```
[the remark this answers is not in the document]
Comment (AsiiaPine) [reply Src/dronecan_application/params.yaml]: Done
image

File: Src/dronecan_application/modules/PWMModule.cpp [hunk 8]
```diff
@@ -0,0 +1,260 @@
...
+        }
+    }
+
+    if (params_error) {
+        module_status = ModuleStatus::MODULE_WARN;
+        if (last_warn_pub_time_ms < HAL_GetTick()) {
+            last_warn_pub_time_ms = HAL_GetTick() + 10000;
+            logger.log_debug("check parameters");
```
[the remark this answers is not in the document]
Comment (AsiiaPine) [reply Src/dronecan_application/modules/PWMModule.cpp:134]: Done in spin_once() function changed initial next_pub_ms value to 5000 ms.
```
    status_pub_timeout_ms = 1;
    static uint32_t next_pub_ms = 5000;

    if (module_status == ModuleStatus::MODULE_OK &&
        crnt_time_ms > next_pub_ms) {
        publish_state();
        next_pub_ms = crnt_time_ms + status_pub_timeout_ms;
    }
```

File: Src/dronecan_application/modules/PWMModule.cpp [hunk 9]
```diff
@@ -0,0 +1,260 @@
...
+        } else {
+            params[i].channel = max_channel;
+            params_error = true;
+        }
+
+        auto min = paramsGetIntegerValue(params_names[i].min);
+        auto max = paramsGetIntegerValue(params_names[i].max);
+        auto def = paramsGetIntegerValue(params_names[i].def);
+        params[i].def = def;
+        if (min == max) {
```
[the remark this answers is not in the document]
Comment (AsiiaPine) [reply Src/dronecan_application/modules/PWMModule.cpp]: Done
```
        auto min = paramsGetIntegerValue(params_names[i].min);
        auto max = paramsGetIntegerValue(params_names[i].max);
        auto def = paramsGetIntegerValue(params_names[i].def);
        params[i].def = def;
        params[i].min = min;
        params[i].max = max;
    }
```
Will the mode be used for hardpoint or smth similar?

[the remark this answers is not in the document]
Comment (PonomarevDA) [reply Src/dronecan_application/modules/PWMModule.cpp]: yes

**Review by PonomarevDA — APPROVED** [review]

**Pull request merged**